tropopause/autoscaling.py

```diff
@@ -0,0 +1,90 @@
+from troposphere import Tags as baseTags
+from troposphere import Base64, Join, Ref
+from troposphere.autoscaling import Tag, Tags
+from troposphere.autoscaling import AutoScalingGroup as upstreamASG
+from troposphere.autoscaling import LaunchConfiguration
+
+
+def AddGenericUserDataRPM(func):
+    """ Add sensible cloudinit into UserData for RPM based systems """
+    def wrapper(*args, **kwargs):
+        if 'UserData' not in kwargs:
+            name = args[1]
+            # pylint: disable=line-too-long
+            cfn_signal = Join(
+                '',
+                [
+                    'INSTANCE=$(curl http://169.254.169.254/latest/meta-data/instance-id 2> /dev/null)',  # noqa
```

Comment:
I can see why you have done this join, but its quite hard to read what these commands are actually doing.

Reply:
perhaps something like:

```python
Join(
    ' && ',
    [
        'INSTANCE=$(curl http://169.254.169.254/latest/meta-data/instance-id 2> /dev/null)',
        'QUERY=AutoScalingInstances[?InstanceId==\\\'$INSTANCE\\\'].[AutoScalingGroupName]',
        'RESOURCE=$(/usr/bin/aws --region={region} autoscaling describe-auto-scaling-instances --query=$QUERY --output=text 2> /dev/null) '
            .format(region=Ref('AWS::Region')),
        '/opt/aws/bin/cfn-signal -e $? --stack {stack} --resource $RESOURCE --region {region}\n'
            .format(stack=Ref('AWS::StackName'), region=Ref('AWS::Region'))
    ]
)
```

the lines are longer, but feels easier to see what it is doing.

Reply:
this goes for the other uses of join too.

https://docs.python.org/3.4/library/string.html#formatspec

Reply:
There's some internal cloudformation / troposphere semantics that makes it not possible to do this, i've tried the format which then embeds the internal python reference as a string, and also tried just inlining { "Ref": "AWS::Foo" }, but this gets escape slashes added.

Reply:
weird, it should be doing almost the same thing... you can add modifies to the fields: `{ref!s}` (i guess calls `.str()`) on the thing, which might help... (possibly what join is doing...)

But presumably calling `join([ 'bla ', Ref('poop'),' bla'])` should be the same as `'bla {poop} bla'.format(poop=Ref('poop'))`